### ISSUE TITLE
Mixed LM optimization GSOC 2015

### ISSUE THREAD

**Issue by kshedden** [statsmodels/statsmodels#2452]
I'm opening this issue for discussion of possible optimization strategies for MixedLM.

The starting point for this work is PR #2363.

First, information gathering:

* Get a rough estimate of where we stand relative to R in terms of total run time.  My experience is that we lag behind LMER by a factor of 3-5 but this is based on very limited examples.

* Get a comparison of the iteration count to convergence for MixedLM and R.  If we are doing more iterations then it is either a problem with the optimizer or we are being too conservative about calling convergence.

* Use the profiler to get a sense of where we are spending time.  I would focus on optimizing the loglikelihood first.  Anything that works there can probably be used in the score function as well.

* Build a set of examples (could be simulated) and a timing script so that when we make changes we know if it really helps.

Ideas for improvement:

* Avoid creating many small temporary arrays by creating re-usable storage for temporaries and using the out keyword parameter, especially in numpy.dot.  One easy place to do this is smw_solver which creates a solver function that is called multiple times (hence storage can be re-used).  Also consider using scipy blas dger to do in-place outer product updates (we have many).  I don't know if there is a good way to use the profiler to see if overhead for creating/destroying temporaries is a significant factor in overall run times.

* Try using sparse matrices for cov_aug and cov_aug_inv (only relevant when variance components are used).  Also sparse matrices could be used for exog_re especially when using a single group to accommodate sparse designs (like random effects ANOVA).

* Find another approach for _gen_dV_dParam : this function helps a lot with readability and extensibility but it might be quite expensive with deeply nested loops and many temporary objects.  Unfortunately it's complex to redesign and still maintain clarity.  The returned arguments are vectors for the regular random effects but are matrices of varying sizes for variance components.

* Avoid re-parsing the formula for each group in from_formula.

* Adjust tuning parameters in the optimizers.  I have noticed in bfgs and lbfgs that sometimes it stops with failed convergence but then when you start again from the ending point it rapidly converges.  Rather than running the optimizer multiple times, can we adjust the convergence criteria to be more strict so that it converges in one call?  I wonder if the issue is that the likelihood is flat so it thinks it is stuck but really needs to just keep going.

@josef-pkt @saketkc

**Comment by saketkc** [comment]
Thanks Kerby!

I need to really dig much deeper. That pushes my timelines by a week or so.
But it kind of will fit in, my post mid-term weeks were mostly buffer ones.

Sorry, I think my contributions so far have been really trivial ones. Still learning.

**Comment by kshedden** [comment]
np.dot(A.ravel(), B.ravel()) should be faster than np.sum(A * B), since the latter creates  a temporary to hold the pointwise matrix product.  I use the latter form in a number of places.  It would be interesting to see how much we gain by switching this.

```
In [6]: %timeit np.dot(A.ravel(), B.ravel())                                                                                                         
100000 loops, best of 3: 14.1 µs per loop

In [7]: %timeit np.sum(A * B)                                                                                                                        
10000 loops, best of 3: 44.9 µs per loop

In [8]: %timeit np.dot(A.T.ravel(), B.ravel())                                                                                                       
10000 loops, best of 3: 28.9 µs per loop

In [9]: %timeit np.dot(A.T.ravel(), B.T.ravel())
10000 loops, best of 3: 42.5 µs per loop
```

**Comment by kshedden** [comment]
take(a, ix) is supposedly faster than a[ix], maybe we can take advantage of this in a few places.

http://wesmckinney.com/blog/numpy-indexing-peculiarities

**Comment by josef-pkt** [comment]
AFAIK, `take` being faster is outdated because of improvements in recent numpy versions

**Comment by kshedden** [comment]
Thanks, for me it actually goes the other way now:

```
In [11]: x = np.random.normal(size=100000)

In [12]: ix = np.random.randint(0, 100000, 1000)

In [13]: %timeit y = x[ix]
100000 loops, best of 3: 3.79 µs per loop

In [14]: %timeit y = np.take(x, ix)
100000 loops, best of 3: 4.85 µs per loop
```

**Comment by saketkc** [comment]
I think they are comparable for larger samples(but `take` still looks slower):
numpy 1.9.2
10000000:
```
x = np.random.normal(size=10000000)

In [11]: ix = np.random.randint(0, 10000000, 10000)

In [12]: %timeit y = np.take(x, ix)
10000 loops, best of 3: 115 µs per loop


In [13]: %timeit y = x[ix]
10000 loops, best of 3: 112 µs per loop
```

10000:
```
In [16]: x = np.random.normal(size=10000)

In [17]: ix = np.random.randint(0, 10000, 1000)

In [18]: %timeit y = np.take(x, ix)
The slowest run took 39.08 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 2.28 µs per loop

In [19]: %timeit y = x[ix]
The slowest run took 82.32 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 1.57 µs per loop
```

**Comment by kshedden** [comment]
There is a ``vex`` defined in score_full that is not used.

**Comment by saketkc** [comment]
>There is a vex defined in score_full that is not used.

Removed it in my old PR: https://github.com/kshedden/statsmodels/pull/1/files#diff-65a46f8f38518560bd427a74998e9d24L1488

**Comment by saketkc** [comment]
This has been really tricky.

The dataset I got hold of using `lme4` gives me a runtime of 30s, while `MixedLM` goes to 12 minutes.
I have tried many alternatives, but they do not seem to be promising so far...

**Comment by kshedden** [comment]
What dataset and what model?  I can give it a try.  You might need to use ``maxiter`` in fit and look at the history to see what is happening.

**Comment by saketkc** [comment]
I picked up example from `lme4-convergence`:
http://nbviewer.ipython.org/gist/saketkc/50b3d999d65fbcc6a134

**Comment by saketkc** [comment]
Profile for a split version of the dataset:
https://gist.github.com/saketkc/dab5f3d8907f62232062

**Comment by kshedden** [comment]
I'll make a few comments but I'm not sure this is a good example to focus on right now.  This is a massive (n ~ 400K) crossed design with no group structure.  It's a good example for the sparse Cholesky approach which we are not following.  I have had surprisingly good luck with crossed designs up to n ~3K but this is two orders of magnitude larger.

Also , your model specification doesn't match the R model you are comparing to, and in addition isn't a valid model specification on its own terms.  It has two random intercepts at the group level (since both of your variance components have an intercept).  In general, you will never want an intercept in a variance component because it gets the same variance parameter as all the other instances of the variance component which usually won't make sense.  Also, you shouldn't have the grouping variable  (item) appearing explicitly in the random effects formula.

For the record, I will try to figure out how to translate the R model specification into our language, but I think we need to start with something more basic.

**Comment by kshedden** [comment]
In the R code they are treating sdiff as quantitative, but you treat it as categorical.  I think it should be quantitative.  Also, the sdiff values are all ~1000 or greater.  There has been some discussion in the lmer lists about variable scaling.  I don't know if lmer autoscales (internally) but we definitely do not.  So I would standardize sdiff.

**Comment by kshedden** [comment]
I don't think we can specify exactly the same model that you fit in R, but we can come close, as below.  However this model cannot be fit on my machine (memory error) as it is ignoring the sparsity structure (specifically in _aex_r, which is 94% sparse in this case, and _aex_r2, which is 98% sparse in this case).

```
modeldata.sdiff = (modeldata.sdiff - modeldata.sdiff.mean()) / modeldata.sdiff.std()

vcf = {"item_int" : "item", "item_sdiff" : "0 + sdiff:C(item)", "subj_int" : "subj", "subj_sdiff" : "0 + sdiff:C(subj)"}
model = sm.MixedLM.from_formula('mean ~ ambiguity * sdiff', groups=np.ones(modeldata.shape[0]), vc_formula=vcf,  data=modeldata)
```

**Comment by saketkc** [comment]
Thanks Kerby. I realized I had the wrong model in `MixedLM`, but went ahead with it since I wanted to identify the bottlenecks.  I also get a Memory error with your suggested model.

Is it okay wot switch to low level BLAS functions from `scipy`? I have not been able to get rid of those loops.

**Comment by kshedden** [comment]
It's not really a bottleneck if the model has redundant parameters.  It can never converge in that case.

Which loops are you referring to?  I don't think the low-level BLAS functions will get rid of any loops.  They do help get rid of some temporary variables.  That may help somewhat with speed but the benefit is likely to be incremental.

I think the most useful next steps would fall under one of the following areas:

1. Information gathering:

* Build a set of test cases, perhaps using vbench or asv, to get a rigorous sense of the performance on a range of test problems.

* Check the number of iterations on our code and in lmer to see if we are spending more time on each iteration versus doing more iterations.

* Do profiling runs (on more conventional problems) to get a sense of where the bottlenecks are.

2. Optimizations

* Add an init keyword to trigger use of sparse matrices.  I think just ``_aex_r`` and ``_aex_r2`` would need to be stored as sparse.  You would then need to add switches to use sparse solvers in ``_smw_solve`` and ``_smw_logdet``.  The goal here would be to enable models for large crossed data like the EEG data you just tried.  The potential payoff is large for certain types of models and zero for other types of models where nothing is sparse.

* Use low level BLAS functions at some key points, especially for outer product updates.  The potential payoff is small to moderate for a broad range of models.

3. Extend the range of models that can be fit:

* Autoregressive errors

* Heteroscedastic errors

4. Add support for various post-estimation tasks

* 2d profile likelihood plots

* ``predict`` is not implemented

* Look to see whatever diagnostic plots are provided by lmer

**Comment by saketkc** [comment]
I was referring to the loops in _gen_dV_Params.
Using small changes such as using `overwrite_a.b`; `check_finite` I was able to get incremental speedup.

My reference would now be Bates book. And my next aim is to support heteroscedasticity.

**Comment by kshedden** [comment]
There are various ways to support heteroscedasticity.  We first should be clear on what approach we are using.

My take on heteroscedasticity for mixed models is that the heteroscedasticity is in the "error terms", not in the factor structure.  That is, if the overall covariance matrix is D + ZPZ', where D is diagonal, then heteroscedasticity means that the diagonal elements of D are not all the same.  This type of heteroscedasticity has nothing to do with P.

This is what I am aware of in terms of options:

1 The user provides fixed variance scaling factors for each endog value within each group.  The diagonal elements of D are proportional to these values.  This is the easiest approach since it doesn't add any parameters to the model.  You would just need to provide a mechanism for the user to pass these values, and use them in smw_solver and smw_logdet as appropriate.

2 User provides group labels for each endog value within each group.  The label indicates which endog values have the same error variance parameter.  These variance parameters are then estimated via ML/REML along with all other parameters.  Basically this gives you heteroscedasticity by group.

3 The variance structure is specified in regression form, via log(D_ii) = c' * t_i, where t_i is a vector of covariates associated with the i^th observation.  The coefficients in c are then estimated by ML/REML along with all other parameters.

In terms of generality, 3 encompases both 2 and 1 as special cases.

A fourth option which I think is not what we really want is something like Stata aweights, which scales the entire row/column of the marginal covariance matrix, something like diag(w^1/2) * (D + Z*psi*Z') * diag(w^1/2).

I prefer option 3, but it is the most ambitious.

**Comment by kshedden** [comment]
Saket found an interesting dataset that is very large and is apparently used to test lmer's ability to handle difficult situations.  It is a data set with around 360K observations, with crossed subjects and items.  In our setup, we have to treat this as a single group which means that the linear algebra is done on very large matrices.  I updated #2363 to allow the design matrices to be passed in sparse form.  They can also be created in sparse form by `from_formula` (but since patsy can't actually create sparse matrices, they are created as dense and then internally converted).  This allows us to fit a crossed model to this data set, although it is very slow (probably close to 20 minutes but I didn't time it).  There are probably a lot more optimizations that can be done with this, but it works, and also still runs fine in the dense case.

The code is below.

```
import statsmodels.api as sm
import pandas as pd
import numpy as np

modeldata = pd.read_csv('https://raw.githubusercontent.com/lme4/lme4-convergence/master/data.tab', delimiter='\t')
modeldata = modeldata[modeldata['win']=='N400']

vcf = {"item" : "1 + C(sdiff):item", "subj" : "1 + C(sdiff):subj"}
model = sm.MixedLM.from_formula('mean ~ ambiguity * sdiff', groups="item", vc_formula=vcf,  data=modeldata)

modeldata.sdiff = (modeldata.sdiff - modeldata.sdiff.mean()) / modeldata.sdiff.std()

vcf = {"item" : "0 + C(item)", "item_slope" : "0 + C(item):sdiff", "subject" : "0 + C(subj)", "subj_sdiff" : "0 + C(subj):sdiff"}
model = sm.MixedLM.from_formula('mean ~ ambiguity * sdiff', groups=np.ones(modeldata.shape[0]),
                                vc_formula=vcf, use_sparse=True, data=modeldata)

result = model.fit(start_params=np.r_[0.1, 0.1, 0.1, 1])
```

**Comment by saketkc** [comment]
Apparently, I was just on this.
I will skip to heteroscedasticity with option 1 for now and incrementally go to option 3 this week.

**Comment by saketkc** [comment]
Kerby, I have some very small changes with minor code cleanup, a pretty moderate speedup with shifting to `scipy` for solve operations and another small change for callbacks and another LR test addition, all depend on #2363. Should I wait till it gets merged?

**Comment by kshedden** [comment]
Thanks Saket.  I am traveling over the next few days but will look this over soon.  I will probably end up cherry picking your changes into #2363.

**Comment by kshedden** [comment]
From an email from @josef-pkt : "If there is time I would like to put extracting a separate class for sparse variance components back on the wishlist. Similar to the comments and discussion in MixedGLM, I would like to see mainly how things fit together and how the (profiled) likelihood and (co)variance parameters can be estimated when the inner part is just a penalized linear model. My guess is that this is largely already in the current MixedLM, but mixed together with the longitudinal model for clusters. My problem is that I know roughly how the penalization (with a normal prior) works because we also use it in several other models, but I don't see how we can calculate and maximize the loglikelihood with respect to the covariance of the prior (random effects distribution) in a memory and computationally efficient way."

The code you recently merged includes some support for sparse structures, specifically in the variance components.  It also includes some other linear algebra optimizations so that we don't form square matrices that are known to be diagonal (e.g. the covariance matrix of variance components).  Therefore, it is possible to fit large crossed models using a single group.  I successfully fit a model with around 100K observations and thousands of random effects (discussed above on this page).  There are certainly more optimizations that can be done, but it works. 

However, the code is not rich enough to support everything, e.g. in R formulas, (1 + x | A) + (1 + z | B) cannot be fit but (1 | A) + (x | A) + (1 | B) + (z  | B) can be fit and (1 + x | A) can be fit.

**Comment by kshedden** [comment]
In MixedLM we do everything through the profile likelihood, not the penalized likelihood.  The Gaussian penalized mixed likelihood should come for free through Greg's project, except that the GLM Gaussian family provides the profiled Gaussian likelihood and derivatives.  I see that it also provides the regular Gaussian likelihood with an explicit scale parameter but I haven't had time to figure out how to trigger that path.

**Comment by josef-pkt** [comment]
What I would like is a simple class with the main structure that could be reused for other models or/and provide a faster special case class.

We also have the penalized likelihood for GLM models in the GAM GSOC project, but that is without any random effects estimation and uses predefined prior covariance with cross-validation or similar for penalization weights. 

MixedLM with variance components would be the simplest case to get the basic structure. Greg's MixedGLM will be mixed together with other parts that are required for MixedGLM and nonlinear integration.

**Comment by kshedden** [comment]
@saketkc if you want to try this, the relevant paper is here:

http://www.stat.wisc.edu/~bates/reports/MixedComp.pdf

You would basically be implementing equations 10 and 27.  

At least equation 10 looks not too difficult.  

If you just do variance components omega will be diagonal.

Remember that many realized variance components will have the same variance parameter, so you will have to unpack params across the diagonal of omega.

One suggestion would be to implement this as a private class called _MixedLM_pls that can then be wrapped later or called from the other MixedLM as an alternative to the profiled approach.
